File: go/pkg/basecamp/subscriptions.go
package basecamp

import (
	"context"
	"fmt"

	"github.com/basecamp/basecamp-sdk/go/pkg/generated"
)

// Subscription represents the subscription state for a recording.
type Subscription struct {
	Subscribed  bool     `json:"subscribed"`
	Count       int      `json:"count"`
	URL         string   `json:"url"`
	Subscribers []Person `json:"subscribers"`
}

// UpdateSubscriptionRequest specifies the parameters for updating subscriptions.
type UpdateSubscriptionRequest struct {
	// Subscriptions is a list of person IDs to subscribe to the recording.
	Subscriptions []int64 `json:"subscriptions,omitempty"`
	// Unsubscriptions is a list of person IDs to unsubscribe from the recording.
	Unsubscriptions []int64 `json:"unsubscriptions,omitempty"`
}

// SubscriptionsService handles subscription operations on recordings.
type SubscriptionsService struct {
	client *Client
}

// NewSubscriptionsService creates a new SubscriptionsService.
func NewSubscriptionsService(client *Client) *SubscriptionsService {
	return &SubscriptionsService{client: client}
}

// Get returns the subscription information for a recording.
// bucketID is the project ID, recordingID is the ID of the recording.
func (s *SubscriptionsService) Get(ctx context.Context, bucketID, recordingID int64) (*Subscription, error) {
	if err := s.client.RequireAccount(); err != nil {
		return nil, err
	}

	resp, err := s.client.gen.GetSubscriptionWithResponse(ctx, bucketID, recordingID)
	if err != nil {
		return nil, err
	}
	if err := checkResponse(resp.HTTPResponse); err != nil {
		return nil, err
	}
	if resp.JSON200 == nil {
		return nil, fmt.Errorf("unexpected empty response")
	}

	subscription := subscriptionFromGenerated(resp.JSON200.Subscription)
	return &subscription, nil
}

// Subscribe subscribes the current user to the recording.
// bucketID is the project ID, recordingID is the ID of the recording.
// Returns the updated subscription information.
func (s *SubscriptionsService) Subscribe(ctx context.Context, bucketID, recordingID int64) (*Subscription, error) {
	if err := s.client.RequireAccount(); err != nil {
		return nil, err
	}

	resp, err := s.client.gen.SubscribeWithResponse(ctx, bucketID, recordingID)
	if err != nil {
		return nil, err
	}
	if err := checkResponse(resp.HTTPResponse); err != nil {
		return nil, err
	}
	if resp.JSON200 == nil {
		return nil, fmt.Errorf("unexpected empty response")
	}

	subscription := subscriptionFromGenerated(resp.JSON200.Subscription)
	return &subscription, nil
}

// Unsubscribe unsubscribes the current user from the recording.
// bucketID is the project ID, recordingID is the ID of the recording.
// Returns nil on success (204 No Content).
func (s *SubscriptionsService) Unsubscribe(ctx context.Context, bucketID, recordingID int64) error {
	if err := s.client.RequireAccount(); err != nil {
		return err
	}

	resp, err := s.client.gen.UnsubscribeWithResponse(ctx, bucketID, recordingID)
	if err != nil {
		return err
	}
	return checkResponse(resp.HTTPResponse)
}

// Update batch modifies subscriptions by adding or removing specific users.
// bucketID is the project ID, recordingID is the ID of the recording.
// Returns the updated subscription information.
func (s *SubscriptionsService) Update(ctx context.Context, bucketID, recordingID int64, req *UpdateSubscriptionRequest) (*Subscription, error) {
	if err := s.client.RequireAccount(); err != nil {
		return nil, err
	}

	if req == nil || (len(req.Subscriptions) == 0 && len(req.Unsubscriptions) == 0) {
		return nil, ErrUsage("at least one of subscriptions or unsubscriptions must be specified")
	}

	body := generated.UpdateSubscriptionJSONRequestBody{
		Subscriptions:   req.Subscriptions,
		Unsubscriptions: req.Unsubscriptions,
	}

	resp, err := s.client.gen.UpdateSubscriptionWithResponse(ctx, bucketID, recordingID, body)
	if err != nil {
		return nil, err
	}
	if err := checkResponse(resp.HTTPResponse); err != nil {
		return nil, err
	}
	if resp.JSON200 == nil {
		return nil, fmt.Errorf("unexpected empty response")
	}

	subscription := subscriptionFromGenerated(resp.JSON200.Subscription)
	return &subscription, nil
}

// subscriptionFromGenerated converts a generated Subscription to our clean type.
func subscriptionFromGenerated(gs generated.Subscription) Subscription {
	s := Subscription{
		Subscribed: gs.Subscribed,
		Count:      int(gs.Count),
		URL:        gs.Url,
	}

	if len(gs.Subscribers) > 0 {
		s.Subscribers = make([]Person, 0, len(gs.Subscribers))
		for _, gp := range gs.Subscribers {
			p := Person{
				Name:         gp.Name,
				EmailAddress: gp.EmailAddress,
				AvatarURL:    gp.AvatarUrl,
				Admin:        gp.Admin,
				Owner:        gp.Owner,
			}
			if gp.Id != nil {
				p.ID = *gp.Id
			}
			s.Subscribers = append(s.Subscribers, p)
		}
	}

	return s
}

File: typescript/src/services/subscriptions.ts
/**
 * Subscriptions service for the Basecamp API.
 *
 * Subscriptions control who receives notifications for a specific recording
 * (like a todo, message, or comment). Users can subscribe or unsubscribe
 * themselves, and you can batch update subscriptions for multiple users.
 *
 * @example
 * ```ts
 * const subscription = await client.subscriptions.get(projectId, recordingId);
 * await client.subscriptions.subscribe(projectId, recordingId);
 * await client.subscriptions.unsubscribe(projectId, recordingId);
 * await client.subscriptions.update(projectId, recordingId, {
 *   subscriptions: [userId1, userId2],
 * });
 * ```
 */

import { BaseService } from "./base.js";
import { Errors } from "../errors.js";

// =============================================================================
// Types
// =============================================================================

/**
 * A person reference (simplified).
 */
export interface PersonRef {
  id: number;
  name: string;
  email_address?: string;
  avatar_url?: string;
  admin?: boolean;
  owner?: boolean;
}

/**
 * Subscription state for a recording.
 */
export interface Subscription {
  /** Whether the current user is subscribed */
  subscribed: boolean;
  /** Number of subscribers */
  count: number;
  /** URL to manage subscriptions */
  url: string;
  /** List of subscribers */
  subscribers: PersonRef[];
}

/**
 * Request to update subscriptions.
 */
export interface UpdateSubscriptionRequest {
  /** Person IDs to subscribe to the recording (optional) */
  subscriptions?: number[];
  /** Person IDs to unsubscribe from the recording (optional) */
  unsubscriptions?: number[];
}

// =============================================================================
// Service
// =============================================================================

/**
 * Service for managing subscriptions in Basecamp.
 */
export class SubscriptionsService extends BaseService {
  /**
   * Gets the subscription information for a recording.
   *
   * @param projectId - The project (bucket) ID
   * @param recordingId - The recording ID
   * @returns The subscription information
   * @throws BasecampError with code "not_found" if recording doesn't exist
   *
   * @example
   * ```ts
   * const subscription = await client.subscriptions.get(projectId, todoId);
   * console.log(subscription.subscribed, subscription.count);
   * subscription.subscribers.forEach(p => console.log(p.name));
   * ```
   */
  async get(projectId: number, recordingId: number): Promise<Subscription> {
    const response = await this.request(
      {
        service: "Subscriptions",
        operation: "Get",
        resourceType: "subscription",
        isMutation: false,
        projectId,
        resourceId: recordingId,
      },
      () =>
        this.client.GET("/buckets/{projectId}/recordings/{recordingId}/subscription.json", {
          params: { path: { projectId, recordingId } },
        })
    );

    return response as unknown as Subscription;
  }

  /**
   * Subscribes the current user to the recording.
   *
   * @param projectId - The project (bucket) ID
   * @param recordingId - The recording ID
   * @returns The updated subscription information
   *
   * @example
   * ```ts
   * const subscription = await client.subscriptions.subscribe(projectId, todoId);
   * console.log("Now subscribed:", subscription.subscribed);
   * ```
   */
  async subscribe(projectId: number, recordingId: number): Promise<Subscription> {
    const response = await this.request(
      {
        service: "Subscriptions",
        operation: "Subscribe",
        resourceType: "subscription",
        isMutation: true,
        projectId,
        resourceId: recordingId,
      },
      () =>
        this.client.POST("/buckets/{projectId}/recordings/{recordingId}/subscription.json", {
          params: { path: { projectId, recordingId } },
        })
    );

    return response as unknown as Subscription;
  }

  /**
   * Unsubscribes the current user from the recording.
   *
   * @param projectId - The project (bucket) ID
   * @param recordingId - The recording ID
   *
   * @example
   * ```ts
   * await client.subscriptions.unsubscribe(projectId, todoId);
   * ```
   */
  async unsubscribe(projectId: number, recordingId: number): Promise<void> {
    await this.request(
      {
        service: "Subscriptions",
        operation: "Unsubscribe",
        resourceType: "subscription",
        isMutation: true,
        projectId,
        resourceId: recordingId,
      },
      () =>
        this.client.DELETE("/buckets/{projectId}/recordings/{recordingId}/subscription.json", {
          params: { path: { projectId, recordingId } },
        })
    );
  }

  /**
   * Batch modifies subscriptions by adding or removing specific users.
   *
   * @param projectId - The project (bucket) ID
   * @param recordingId - The recording ID
   * @param req - Subscription update parameters
   * @returns The updated subscription information
   * @throws BasecampError with code "validation" if neither subscriptions nor unsubscriptions is provided
   *
   * @example
   * ```ts
   * const subscription = await client.subscriptions.update(projectId, todoId, {
   *   subscriptions: [userId1, userId2],
   *   unsubscriptions: [userId3],
   * });
   * ```
   */
  async update(
    projectId: number,
    recordingId: number,
    req: UpdateSubscriptionRequest
  ): Promise<Subscription> {
    if (
      (!req.subscriptions || req.subscriptions.length === 0) &&
      (!req.unsubscriptions || req.unsubscriptions.length === 0)
    ) {
      throw Errors.validation(
        "At least one of subscriptions or unsubscriptions must be specified"
      );
    }

    const response = await this.request(
      {
        service: "Subscriptions",
        operation: "Update",
        resourceType: "subscription",
        isMutation: true,
        projectId,
        resourceId: recordingId,
      },
      () =>
        this.client.PUT("/buckets/{projectId}/recordings/{recordingId}/subscription.json", {
          params: { path: { projectId, recordingId } },
          body: {
            subscriptions: req.subscriptions,
            unsubscriptions: req.unsubscriptions,
          },
        })
    );

    return response as unknown as Subscription;
  }
}
